### PR TITLE
Closes #64

### LINKED ISSUE #64: Create a dataset loader for CodiEsp
From https://zenodo.org/record/3837305#.YL46cfdfjMU

### PR DESCRIPTION
**Pull request by mapama247**
- **Name:** CODIESP
- **Description:** Collection of 1,000 manually selected clinical case studies in Spanish.
- **Paper:** [link to the dataset paper](http://ceur-ws.org/Vol-2696/paper_263.pdf)
- **Data:** [link to the online home of the dataset](https://temu.bsc.es/codiesp)

### Checkbox

- [x] Confirm that this PR is linked to the dataset issue.
- [x] Create the dataloader script `biodatasets/my_dataset/my_dataset.py` (please use only lowercase and underscore for dataset naming).
- [x] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_BIGBIO_VERSION` variables.
- [x] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [x] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `BigBioConfig` for the source schema and one for a bigbio schema.
- [x] Confirm dataloader script works with `datasets.load_dataset` function.
- [x] Confirm that your dataloader script passes the test suite run with `python -m tests.test_bigbio biodatasets/my_dataset/my_dataset.py`.
- [x] If my dataset is local, I have provided an output of the unit-tests in the PR (please copy paste). This is OPTIONAL for public datasets, as we can test these without access to the data files.

### CODE REVIEW

File: biodatasets/codiesp/codiesp.py [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+
+class CodiespDataset(datasets.GeneratorBasedBuilder):
+    """Collection of 1,000 manually selected clinical case studies in Spanish."""
+
+    SOURCE_VERSION = datasets.Version(_SOURCE_VERSION)
+    BIGBIO_VERSION = datasets.Version(_BIGBIO_VERSION)
+
+    BUILDER_CONFIGS = [
+        BigBioConfig(
+            name="codiesp_source",
```
Comment (sg-wbi) [inline biodatasets/codiesp/codiesp.py:101]: Would it be possible to have the different subtask for the `source` schema too?

File: biodatasets/codiesp/codiesp.py [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+            version=SOURCE_VERSION,
+            description="CodiEsp source schema",
+            schema="source",
+            subset_id="codiesp",
+        ),
+        BigBioConfig(
```
Comment (sg-wbi) [inline biodatasets/codiesp/codiesp.py:107]: In many datasets I saw the subtrack is usually in this format`codiesp_P_bigbio_text`. Would you mind change it to this?

**Review by sg-wbi — COMMENTED** [review]
Thank you for your contribution @mapama247 ! Sorry for the delay. Unfortunately I could not run the tests because the download link does not seem to work. I tried in two different occasion/networks but still nothing. Is there any chance we can get it somewhere else? Or maybe you can create a github repo with the data if you still have it. This is just to double check that the script works fine!

File: biodatasets/codiesp/codiesp.py (hunk 2 above)
Comment (mapama247) [reply biodatasets/codiesp/codiesp.py:107]: sure! done

File: biodatasets/codiesp/codiesp.py (hunk 1 above)
Comment (sg-wbi) [reply biodatasets/codiesp/codiesp.py:101]: That would be great thanks! This way we have `codiesp_<subtask>_source` and `codiesp_<subtask>_bigbio_text`

**Comment by sg-wbi** [comment]
I can now download the dataset w/o problems! Who knows what was the problem... thanks for double checking though!

**Comment by mapama247** [comment]
> I can now download the dataset w/o problems! Who knows what was the problem... thanks for double checking though!

Glad to hear that! must have something to do with zenodo... I have just updated the script with a different source config for each task, let me know if everything is fine now.

File: biodatasets/codiesp/codiesp.py [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+@inproceedings{
+  title={Overview of automatic clinical coding: annotations, guidelines, and \
+  solutions for non-english clinical cases at codiesp track of CLEF eHealth 2020},
+  author={Miranda-Escalada, Antonio and Gonzalez-Agirre, Aitor and Armengol-Estap{\'e}, Jordi and Krallinger, Martin},
+  booktitle={Working Notes of Conference and Labs of the Evaluation (CLEF) Forum. CEUR Workshop Proceedings},
+  year={2020}
+}
+"""
+
+_DATASETNAME = "cantemist"
```
Comment (sg-wbi) [inline biodatasets/codiesp/codiesp.py:48]: Why `cantemist` and not `codiesp`?

File: biodatasets/codiesp/codiesp.py [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+CodiEsp-P: The Procedure Coding sub-task, which requires automatic ICD10-PCS [CIE10-Procedimiento] code assignment.
+
+CodiEsp-X: The Explainable AI exploratory sub-task, which requires to submit the reference to the predicted codes \
+(both ICD10-CM and ICD10-PCS). The goal of this novel task was not only to predict the correct codes but also to \
+present the reference in the text that supports the code predictions.
+
+For further information, please visit https://temu.bsc.es/codiesp or send an email to encargo-pln-life@bsc.es
+"""
+
+_HOMEPAGE = "https://temu.bsc.es/codiesp/index.php/2019/09/19/datasets/"
```
Comment (sg-wbi) [inline biodatasets/codiesp/codiesp.py:78]: I would change this to "https://temu.bsc.es/codiesp/". It has a nicer overview of the tasks.

File: biodatasets/codiesp/codiesp.py [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+
+_LICENSE = "Creative Commons Attribution 4.0 International"
+
+_URLS = {
+    "codiesp": "https://zenodo.org/record/3837305/files/codiesp.zip?download=1",
+}
+
+_SUPPORTED_TASKS = [Tasks.TEXT_CLASSIFICATION]
```
Comment (sg-wbi) [inline biodatasets/codiesp/codiesp.py:86]: We should add `Tasks.NAMED_ENTITY_RECOGNITION` and `Tasks.NAMED_ENTITY_DISAMBIGUATION` for `codiesp_X`

File: biodatasets/codiesp/codiesp.py [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+                yield guid, example
+
+        elif self.config.name == "codiesp_X_bigbio_kb":
+            df = pd.read_csv(paths[self.config.subset_id], sep="\t", header=None)
+
+            example = {}
+            for guid, row in df.iterrows():
+                file, label, code, text, spans = row[0], row[1], row[2], row[3], row[4]
+                appearances = spans.split(";")
+                spans = []
```
Comment (sg-wbi) [inline biodatasets/codiesp/codiesp.py:258]: The way I understood it from [here (see CodiEsp Explainable AI exploratory) ](https://temu.bsc.es/codiesp/) is that the spans are "words in the text" that explain (are the reason why) a certain ICD10-CM/ICD10-PCS code was added.

To me this sounds like a `Tasks.NAMED_ENTITY_DISAMBIGUATION`. The dataset has lines in the form:
```bash
S0004-06142005000700014-1   PROCEDIMIENTO   bw03zzz Rx tórax   2163 2171
S0004-06142005000700014-1   DIAGNOSTICO n44.8   teste derecho aumentado de tamaño  1343 1376
```
So for instance I would convert the first line in `bigbio_kb` schema as:
```python
        "entities": [
            {
                "id": <unique id>,
                "type": "PROCEDIMIENTO", 
                "text": "Rx tórax",
                "offsets": [(2163,2171)],
                "normalized": [
                    {
                        "db_name": "ICD10-PCS",
                        "db_id": bw03zzz,
                    }
                ],
            }
        ],
```
Does it make sense?

**Review by sg-wbi — COMMENTED** [review]
@mapama247 thank you for applying the changes! Now that I could finally take a look at the dataset I have some more in depth comments. Could you please check them out? Thank you!

File: biodatasets/codiesp/codiesp.py (hunk 3 above)
Comment (mapama247) [reply biodatasets/codiesp/codiesp.py]: ups, I mixed things (cantemist is another dataset I was working on)

File: biodatasets/codiesp/codiesp.py (hunk 4 above)
Comment (mapama247) [reply biodatasets/codiesp/codiesp.py]: done!

File: biodatasets/codiesp/codiesp.py (hunk 5 above)
Comment (mapama247) [reply biodatasets/codiesp/codiesp.py]: done!

**Comment by mapama247** [comment]
> Ah! One more thing: [here (see bottom of the page )](https://temu.bsc.es/codiesp/index.php/2019/09/19/datasets/) they provide an "Additional dataset" to extend train/dev splits. Could you please add these as well? Or maybe it is worth to create a new Issue and a separate dataloader script? What do you think?

I'm not sure... I guess that if we add it in the same script this would simply require an additional BigBioConfig, what schema should I use and in what fields would you want me to store the data (title, abstract, pmid, mesh...)

File: biodatasets/codiesp/codiesp.py [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+                    "relations": []
+                }
+
+                for idx, d in enumerate(data):
+                    example["entities"].append({
+                                            "id": str(guid)+str(idx),
+                                            "type": d["label"],
+                                            "text": [d["text"]],
+                                            "offsets": d["spans"],
+                                            "normalized": [{"db_name": "ICD10-PCS",
```
Comment (sg-wbi) [inline biodatasets/codiesp/codiesp.py:284]: This `"db_name"` should be `ICD10-PCS` if the entity is `PROCEDIMIENTO`, but it should be `ICD10-CM` if the entity type is `DIAGNOSTICO`

**Comment by sg-wbi** [comment]
Regarding the additional split:
The data [here](https://zenodo.org/record/3606662#.YmJkz_exVqg) it's in the following format:

```json
{"articles": [{"title": "Influencia: un problema Pol\u00edtico-Terap\u00e9utico en la Genealog\u00eda del Psicoan\u00e1lisis", "pmid": "biblio-994981", "abstractText": "El art\u00edculo tiene por objectivo trazar una genealog\u00eda possible de la t\u00e9cnica psicoanal\u00edtica marcando la importancia de las dimenciones hist\u00f3ricas, pol\u00edtica y sociales en la construcci\u00f3n de la psicoan\u00e1lisis. A partir del fen\u00f4meno del magnetismo animal, propuesto por Franz Anton Mesmer, en siglo XVIII, pasando por la hipnosis y por la sugesti\u00f3n, hasta encuentrarmos la transferencia psicoanal\u00edtica, intentamos remontar la influencia como pressupuesto \u00e9tico-pol\u00edtico que atraviesa y possibilita todas esas practicas.(AU)", "Mesh": [{"Code": "D006801", "Word": "Humanos"}, {"Code": "D007989", "Word": "Libido", "CIE": ["R68.82"]}, {"Code": "D011572", "Word": "Psicoan\u00e1lisis"}, {"Code": "D006990", "Word": "Hipnosis", "CIE": ["GZFZZZZ"]}, {"Code": "D011057", "Word": "Pol\u00edtica"}]}
```
This would turn out to be a `Tasks.TEXT_CLASSIFICATION` with examples in the form:
```python
	"id": "<unique_id>",
	"document_id": biblio-994981 , # "pmid" key
	"text": "TEXT" , # "title" + "abstractText" keys
	"labels": ["D006801", ...], # "Mesh" key
```
I would keep this data attached to this dataloader since it was designed to be an extra training split. A key aspect here is that the labels sometimes contain both a `MeSH` ID (e.g. D006801) and an `CIE` ID (e.g. R68.82) or only a `MeSH` ID. So, the way I see it this should result in two (four counting source) extra configs with a train split:

```python
        BigBioConfig(
            name="codiesp_extra_mesh_bigbio_text",
            version=SOURCE_VERSION,
            description="Abstracts from Lilacs and Ibecs with MESH Codes",
            schema="bigbio",
            subset_id="codiesp_extra_mesh",
        ),
    BigBioConfig(
            name="codiesp_extra_cie_bigbio_text",
            version=SOURCE_VERSION,
            description="Abstracts from Lilacs and Ibecs with CIE10 Codes",
            schema="bigbio",
            subset_id="codiesp_extra_cie",
        ),
```
Does it make sense?

File: biodatasets/codiesp/codiesp.py (hunk 7 above)
Comment (mapama247) [reply biodatasets/codiesp/codiesp.py]: well seen,  this has been fixed now

**Comment by mapama247** [comment]
> Does it make sense?

Sure! I followed your suggestions and added four extra configs to hold the abstracts dataset :)

**Comment by sg-wbi** [comment]
Great work @mapama247 ! The dataset is now complete! Thank you very much for this contribution! 
I'll wait to merge this since there seems to be an issue w/ testing, i.e. if I run
```bash
python -m tests.test_bigbio biodatasets/codiesp/codiesp.py --subset_id codiesp_p
```
I get this error
```bash
Traceback (most recent call last):
  File "/home/ele/Desktop/projects/tm/bigbio/tests/test_bigbio.py", line 187, in setUp
    self.datasets_bigbio[schema] = datasets.load_dataset(
  File "/home/ele/.venv/bigbio/lib/python3.9/site-packages/datasets/load.py", line 1675, in load_dataset
    builder_instance = load_dataset_builder(
  File "/home/ele/.venv/bigbio/lib/python3.9/site-packages/datasets/load.py", line 1533, in load_dataset_builder
    builder_instance: DatasetBuilder = builder_cls(
  File "/home/ele/.venv/bigbio/lib/python3.9/site-packages/datasets/builder.py", line 1020, in __init__
    super().__init__(*args, **kwargs)
  File "/home/ele/.venv/bigbio/lib/python3.9/site-packages/datasets/builder.py", line 258, in __init__
    self.config, self.config_id = self._create_builder_config(
  File "/home/ele/.venv/bigbio/lib/python3.9/site-packages/datasets/builder.py", line 349, in _create_builder_config
    raise ValueError(f"BuilderConfig {name} not found. Available: {list(self.builder_configs.keys())}")
ValueError: BuilderConfig codiesp_p_bigbio_kb not found. Available: ['codiesp_d_source', 'codiesp_p_source', 'codiesp_x_source', 'codiesp_extra_mesh_source', 'codiesp_extra_cie_source', 'codiesp_d_bigbio_text', 'codiesp_p_bigbio_text', 'codiesp_x_bigbio_kb', 'codiesp_extra_mesh_bigbio_text', 'codiesp_extra_cie_bigbio_text']
```
However, to me it's the error being wrong, i.e. we are not supposed to run tests for "codiesp_p_bigbio_kb" if the `subset_id` only supports `bigbio_text`.

As soon as we figure this out (see #398) I'll merge this.

**Comment by hakunanatasha** [comment]
> Great work @mapama247 ! The dataset is now complete! Thank you very much for this contribution! I'll wait to merge this since there seems to be an issue w/ testing, i.e. if I run
> 
> ```shell
> python -m tests.test_bigbio biodatasets/codiesp/codiesp.py --subset_id codiesp_p
> ```
> 
> I get this error
> 
> ```shell
> Traceback (most recent call last):
>   File "/home/ele/Desktop/projects/tm/bigbio/tests/test_bigbio.py", line 187, in setUp
>     self.datasets_bigbio[schema] = datasets.load_dataset(
>   File "/home/ele/.venv/bigbio/lib/python3.9/site-packages/datasets/load.py", line 1675, in load_dataset
>     builder_instance = load_dataset_builder(
>   File "/home/ele/.venv/bigbio/lib/python3.9/site-packages/datasets/load.py", line 1533, in load_dataset_builder
>     builder_instance: DatasetBuilder = builder_cls(
>   File "/home/ele/.venv/bigbio/lib/python3.9/site-packages/datasets/builder.py", line 1020, in __init__
>     super().__init__(*args, **kwargs)
>   File "/home/ele/.venv/bigbio/lib/python3.9/site-packages/datasets/builder.py", line 258, in __init__
>     self.config, self.config_id = self._create_builder_config(
>   File "/home/ele/.venv/bigbio/lib/python3.9/site-packages/datasets/builder.py", line 349, in _create_builder_config
>     raise ValueError(f"BuilderConfig {name} not found. Available: {list(self.builder_configs.keys())}")
> ValueError: BuilderConfig codiesp_p_bigbio_kb not found. Available: ['codiesp_d_source', 'codiesp_p_source', 'codiesp_x_source', 'codiesp_extra_mesh_source', 'codiesp_extra_cie_source', 'codiesp_d_bigbio_text', 'codiesp_p_bigbio_text', 'codiesp_x_bigbio_kb', 'codiesp_extra_mesh_bigbio_text', 'codiesp_extra_cie_bigbio_text']
> ```
> 
> However, to me it's the error being wrong, i.e. we are not supposed to run tests for "codiesp_p_bigbio_kb" if the `subset_id` only supports `bigbio_text`.
> 
> As soon as we figure this out (see #398) I'll merge this.

@sg-wbi the issue is that we inspect the tasks -the dataset tasks require relation extraction and NER; there is no 'text' task which is why it's throwing this error. Either the name of the view is incorrect (should be _kb) or there is a task missing in `_SUPPORTED_TASKS`

**Comment by sg-wbi** [comment]
Hey @hakunanatasha thank you for taking a look at this. 

> Either the name of the view is incorrect (should be _kb) or there is a task missing in _SUPPORTED_TASKS

I think that the name is correct, because the "split" `codiesp_p` is a text classification task and in the `_SUPPORTED_TASKS` there is a `Tasks.TEXT_CLASSIFICATION`. The way I see it we should run tests based on the `schema` provided by the `subset_id` or have a bypass arg. I'm seeing the same issue [here](https://github.com/bigscience-workshop/biomedical/pull/488#issuecomment-1110340963)

**Comment by hakunanatasha** [comment]
@sg-wbi great call - I will fix this in the refactor of the unit-tests.

**Comment by hakunanatasha** [comment]
@sg-wbi can we just run --subset_id and specifically temporarily remove the other _SUPPORTED_TASKS that relate to the kb dataset? This will be a hack to check if the unit-tests work and workaround the fact that it's failing on this unique case even if the structure is correct.

**Comment by hakunanatasha** [comment]
@sg-wbi the new unit-tests take care of this consideration. I will merge.

**Review by hakunanatasha — APPROVED** [review]

**Pull request merged**